File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mp = new HashMap<>();
        int l=0;
        int ans = 0;
        for (int r = 0; r < s.length(); r++) {
            char currentChar = s.charAt(r);
            if (mp.containsKey(currentChar) && mp.get(currentChar) >= l) {
                l = mp.get(currentChar) + 1;
            }
            mp.put(currentChar, r);
            ans = Math.max(ans, r - l + 1);
        }
        return ans;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d = {}
        l = r = maxlen = 0
        while(r<len(s)):
            if(s[r] in d):
                if(d[s[r]]>=l):
                    l = d[s[r]] + 1
            length = r - l + 1
            maxlen = max(length, maxlen)
            d[s[r]] = r
            r += 1
        return maxlen